File: 20-valid-parentheses/valid-parentheses.java
// Method 1: Without HashMap
/*
We only pop when we see a closer and it matches the expected opener on top of the stack.
Any mismatch or early closer with an empty stack returns false.
At the end, the stack must be empty (no unmatched openers left).
*/
class Solution {
    public boolean isValid(String s) {
        Deque<Character> st = new ArrayDeque<>();
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            switch (c) {
                case '(': case '{': case '[':
                    st.push(c);
                    break;
                case ')':
                    if (st.isEmpty() || st.pop() != '(') return false;
                    break;
                case '}':
                    if (st.isEmpty() || st.pop() != '{') return false;
                    break;
                case ']':
                    if (st.isEmpty() || st.pop() != '[') return false;
                    break;
                default:
                    return false; // if inputs can contain other chars
            }
        }
        return st.isEmpty();
    }
}



// Method 2: Using HashMap
// class Solution {
//     public boolean isValid(String s) {
//         Deque<Character> st = new ArrayDeque<>();
//         Map<Character, Character> need = Map.of(
//             ')', '(',
//             '}', '{',
//             ']', '['
//         );

//         for (int i = 0; i < s.length(); i++) {
//             char c = s.charAt(i);

//             if (need.containsKey(c)) {            // c is a closing bracket
//                 if (st.isEmpty() || st.pop() != need.get(c)) {
//                     return false;                 // mismatch or no opener
//                 }
//             } else {                               // c is an opening bracket
//                 st.push(c);
//             }
//         }
//         return st.isEmpty();                       // no unmatched openers
//     }
// }

File: 20-valid-parentheses/valid-parentheses.cpp
// Method 1: Stack approach
/*
*/
class Solution {
public:
    bool isValid(string s) {
        stack<char> stack;

        for(char c: s){
            if(c == '(' || c == '{' || c == '['){
                stack.push(c);
            }else if(c == ')'){
                if(stack.empty() || stack.top() != '('){
                    return false;
                }
                stack.pop();
            }else if(c == '}'){
                if(stack.empty() || stack.top() != '{'){
                    return false;
                }
                stack.pop();
            }else if(c == ']'){
                if(stack.empty() || stack.top() != '['){
                    return false;
                }
                stack.pop();
            }
        }

        return stack.empty();
    }
};

// Synced seamlessly with LeetHub Pro
// Pro features: https://bit.ly/leethubpro | Free version: https://bit.ly/leethubv4
// Get it here: https://chromewebstore.google.com/detail/bcilpkkbokcopmabingnndookdogmbna